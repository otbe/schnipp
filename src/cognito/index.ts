import { CognitoUserPoolTriggerEvent } from 'aws-lambda';
import { IHandler } from '../';

export type CognitoUserPoolHandler = IHandler<
  CognitoUserPoolTriggerEvent,
  CognitoUserPoolTriggerEvent
>;
