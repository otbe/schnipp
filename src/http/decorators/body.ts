import { createMethodParameterDecorator } from '../factories/createMethodParamDecorator';

export const Body = () =>
  createMethodParameterDecorator(e => JSON.parse(e.body || '{}'));
