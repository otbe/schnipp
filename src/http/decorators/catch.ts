const CATCH = Symbol('catch');

export const Catch = (...exceptions: any[]): ClassDecorator => (
  target: any
) => {
  Reflect.defineMetadata(CATCH, exceptions, target.constructor);
};

export const getCaughtExceptions = (catcher: any) =>
  Reflect.getMetadata(CATCH, catcher.constructor) || [];
