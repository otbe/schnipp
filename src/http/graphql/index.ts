export { GraphQLController } from './GraphQLController';
export { Query } from './decorators/query';
export { Mutation } from './decorators/mutation';
export { ResolveField } from './decorators/resolveField';
export { ResolveType } from './decorators/resolveType';
export { Header } from './decorators/header';
export { Meta } from './decorators/meta';
export { Use } from './decorators/use';
