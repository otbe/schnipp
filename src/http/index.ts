import {
  APIGatewayEvent,
  APIGatewayProxyResult,
  CustomAuthorizerEvent,
  CustomAuthorizerResult
} from 'aws-lambda';
import { IHandler } from '..';

export type APIGatewayHandler = IHandler<
  APIGatewayEvent,
  APIGatewayProxyResult
>;

export type AuthorizeHandler = IHandler<
  CustomAuthorizerEvent,
  CustomAuthorizerResult
>;

export { AuthPolicy, HttpVerb, Effect } from './utils/AuthPolicy';
export { createResponse } from './utils/createResponse';
export { MetaData } from './utils/MetaData';
export {
  DecoratedExceptionFilter,
  ExceptionFilter
} from './utils/ExceptionFilter';
export { Guard } from './utils/Guard';
export { DefaultExecutionContext } from './utils/ExecutionContext';
export { Catch } from './decorators/catch';
