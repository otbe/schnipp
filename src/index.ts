import { Context, Handler } from 'aws-lambda';
import { ContainedType, getFromContainer } from './container';

export type IHandler<E, R = void> = {
  handle(event: E, context: Context): Promise<R> | R;
};

export const createHandler = <E, R>(
  handler: ContainedType<IHandler<E, R>>
): Handler<E, R> => async (event: E, context: Context) => {
  context.callbackWaitsForEmptyEventLoop = false;

  const handlerInstance = getFromContainer(handler);
  return await handlerInstance.handle(event, context);
};

export { useContainer } from './container';
