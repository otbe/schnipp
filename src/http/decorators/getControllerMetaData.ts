import { DecoratedExceptionFilter, Guard, MetaData } from '..';
import { ContainedType } from '../../container';
import { GraphQLController } from '../graphql';
import { RestController } from '../rest';

export const CONTROLLER = Symbol('controller');

export type ControllerBaseData = {
  headers: { [key: string]: string };
  filters: Array<DecoratedExceptionFilter>;
  guards: Array<ContainedType<Guard>>;
  metaData: MetaData;
};

export type GraphQLControllerData = ControllerBaseData;

export type RestControllerData = ControllerBaseData & {
  basePath?: string;
};

export type ControllerData = GraphQLControllerData | RestControllerData;

export const getControllerMetaData = (
  controller: typeof RestController | typeof GraphQLController
) => {
  let data: ControllerData | null = Reflect.getMetadata(CONTROLLER, controller);

  if (data == null) {
    data = {
      filters: [],
      guards: [],
      metaData: {},
      headers: {}
    };

    Reflect.defineMetadata(CONTROLLER, data, controller);
  }

  return data;
};
