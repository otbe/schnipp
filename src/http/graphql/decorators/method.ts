import { getResolverMetaData } from './resolver';
import { Newable } from 'simple-ts-di';
import { DecoratedExceptionFilter, Guard, MetaData } from '../..';
import { DocumentNode } from 'graphql';

export const METHOD = Symbol('method');

export type Fields = {
  [fieldName: string]: ResolverMethodMeta<any>;
};

export type ResolverMeta = {
  resolverMap: { [typeName: string]: Fields };
};

export type ResolverMethodMeta<T> = {
  clazz?: Newable<T>;
  methodName: string;
  filters: Array<DecoratedExceptionFilter>;
  guards: Array<Newable<Guard>>;
  metaData: MetaData;
};

export const getMethodMetaData = (target: any, methodName: string) => {
  let data: ResolverMethodMeta<any> | null = Reflect.getMetadata(
    METHOD,
    target,
    methodName
  );

  if (data == null) {
    data = {
      methodName,
      filters: [],
      guards: [],
      metaData: {}
    };

    Reflect.defineMetadata(METHOD, data, target, methodName);
  }

  return data;
};

export const createMethodDecorator = (typeName: string) => (
  fieldName: string
): MethodDecorator => <T>(
  resolver: any,
  propertyKey: string | symbol,
  descriptor: TypedPropertyDescriptor<T>
) => {
  const resolverMeta = getResolverMetaData(resolver.constructor);
  resolverMeta.resolverMap[typeName] = resolverMeta.resolverMap[typeName] || {};

  const methodMetaData = getMethodMetaData(resolver, propertyKey.toString());

  methodMetaData.clazz = resolver.constructor;
  resolverMeta.resolverMap[typeName][fieldName] = methodMetaData;

  return descriptor;
};
