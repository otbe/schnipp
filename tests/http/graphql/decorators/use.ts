import 'reflect-metadata';
import { createResponse, ExceptionFilter, Guard } from '../../../../src/http';
import { getControllerMetaData } from '../../../../src/http/decorators/getControllerMetaData';
import { GraphQLController, Query, Use } from '../../../../src/http/graphql';
import { getMethodMetaData } from '../../../../src/http/graphql/decorators/method';

describe('@Use', () => {
  it('should add guards and filters for methods and instances of GraphQLController', () => {
    class TestGuard implements Guard {
      canActivate() {
        return true;
      }
    }

    class Filter implements ExceptionFilter {
      catch() {
        return createResponse(200);
      }
    }

    class TestResolver {
      @Query('foo')
      @Use(TestGuard, Filter)
      foo() {}
    }

    @Use(TestGuard, Filter)
    class Test extends GraphQLController {
      getResolvers() {
        return [TestResolver];
      }
    }

    expect(getControllerMetaData(Test)).toMatchSnapshot();
    expect(getMethodMetaData(new TestResolver(), 'foo')).toMatchSnapshot();
  });

  it('should throw on wrong usage', () => {
    expect(() => Use()()).toThrow();
    expect(() => Use()(1, 2, 3, 4)).toThrow();
  });
});
