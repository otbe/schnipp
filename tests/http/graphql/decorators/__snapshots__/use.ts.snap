// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@Use should add guards and filters for methods and instances of GraphQLController 1`] = `
Object {
  "filters": Array [
    Object {
      "exceptions": Array [],
      "filter": [Function],
    },
  ],
  "guards": Array [
    [Function],
  ],
  "headers": Object {},
  "metaData": Object {},
}
`;

exports[`@Use should add guards and filters for methods and instances of GraphQLController 2`] = `
Object {
  "clazz": [Function],
  "filters": Array [
    Object {
      "exceptions": Array [],
      "filter": [Function],
    },
  ],
  "guards": Array [
    [Function],
  ],
  "metaData": Object {},
  "methodName": "foo",
}
`;
